### PR TITLE
[AutoDiff] First cut of coroutines differentiation

### LINKED ISSUE #54404: [SR-14115] SIL: derivative type calculation for coroutines
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-14115      |
   |Radar             | None         |
   |Original Reporter | @dan-zheng      |
   |Type              | Sub-task    |

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Compiler    |
   |Labels            | Sub-task, AutoDiff        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: fe00df47fa9d2ad8d02ae88725898756

  </details>

**Parent-Task**:  
* [SR-14113](https://bugs.swift.org/browse/SR-14113) Support `_read` and `_modify` accessor differentiation  

**Issue Description:**

SIL has dedicated coroutine function types: <https://github.com/apple/swift/blob/master/docs/SIL.rst#coroutine-types>

Figure out derivative type calculation for coroutines.

``` java
// Array.subscript.modify example
Original function type: $@yield_once @convention(method) <τ_0_0> (Int, @inout Array<τ_0_0>) -> @yields @inout τ_0_0 
Derivative function type: ???
```

### PR DESCRIPTION
**Pull request by asl**
This PR implements first set of changes required to support autodiff for coroutines. It mostly targeted to `_modify` accessors in standard library (and beyond), but overall implementation is quite generic.

There are some specifics of implementation and known limitations:
 - Only `@yield_once` coroutines are naturally supported
 - VJP is a coroutine itself: it yields the results *and* returns a pullback closure as a normal return. This allows us to capture values produced in resume part of a coroutine (this is required for defers and other cleanups / commits)
 - Pullback is a coroutine, we assume that coroutine cannot abort and therefore we execute the original coroutine in reverse from return via yield and then back to the entry
 - It seems there is no semantically sane way to support `_read` coroutines (as we will need to "accept" adjoints via yields), therefore only coroutines with inout yields are supported (`_modify` accessors). Pullbacks of such coroutines take adjoint buffer as input argument, yield this buffer (to accumulate adjoint values in the caller) and finally return the adjoints indirectly.
 - Coroutines (as opposed to normal functions) are not first-class values: there is no AST type for them, one cannot e.g. store them into tuples, etc. So, everywhere where AST type is required, we have to hack around.
 - As there is no AST type for coroutines, there is no way one could register custom derivative for coroutines. So far only compiler-produced derivatives are supported
 - There are lots of common things wrt normal function apply's, but still there are subtle but important differences. I tried to organize the code to enable code reuse, still it was not always possible, so some code duplication could be seen
 - The order of how pullback closures are produced in VJP is a bit different: for normal apply's VJP produces both value and pullback closure via a single nested VJP apply. This is not so anymore with coroutine VJP's: yielded values are produced at `begin_apply` site and pullback closure is available only from `end_apply`, so we need to track the order in which pullbacks are produced (and arrange consumption of the values accordingly – effectively delay them)
 - On the way some complementary changes were required in e.g. mangler / demangler

This patch covers the generation of derivatives up to SIL level, however, it is not enough as codegen of `partial_apply` of a coroutine is completely broken. The fix for this will be submitted separately as it is not directly autodiff-related.

Some diagnostics were contributed by @asavonic

### CODE REVIEW

**Comment by asl** [comment]
Tagging @BradLarson

**Comment by asl** [comment]
@swift-ci please test

**Comment by asl** [comment]
@swift-ci please test windows platform

**Review by rxwei — COMMENTED** [review]
Should the changes to `begin_apply` and `end_apply` be in their own PR before the differentiation aspects?

**Comment by asl** [comment]
> Should the changes to `begin_apply` and `end_apply` be in their own PR before the differentiation aspects?

They are and this is https://github.com/apple/swift/pull/71645. This PR will be rebased after https://github.com/apple/swift/pull/71645 will be merged.

To be reviewed are 3aba19bf095c2435b8ebc19affcf76a7595c91f2 and everything after

**Comment by asl** [comment]
Another pre-requisite is `partial_apply` patch for coroutines. It is submitted  in https://github.com/apple/swift/pull/71653 As I mentioned, w/o #71653 we will generate proper SIL, but will be unable to proceed further down to real code generation and execution

**Comment by asl** [comment]
@rxwei Rebased and ready for review. Please take a look when you will have a chance.

**Comment by asl** [comment]
@swift-ci please test

File: include/swift/AST/DiagnosticsSIL.def [hunk 1]
```diff
@@ -604,6 +604,8 @@ NOTE(autodiff_cannot_differentiate_through_multiple_results,none,
      "cannot differentiate through multiple results", ())
 NOTE(autodiff_cannot_differentiate_through_inout_arguments,none,
      "cannot differentiate through 'inout' arguments", ())
+NOTE(autodiff_cannot_differentiate_through_direct_yield,none,
+     "cannot differentiate through a direct yield result", ())
```
Comment (rxwei) [inline include/swift/AST/DiagnosticsSIL.def:608]: Perhaps it's clearer to say "read accessor" instead?
```suggestion
     "cannot differentiate through a '_read' accessor", ())
```

**Review by rxwei — APPROVED** [review]

File: test/AutoDiff/SILOptimizer/activity_analysis.swift [hunk 2]
```diff
@@ -571,53 +571,81 @@ func testTryApply(_ x: Float) -> Float {
...
 }
+
+
 // expected-error @+1 {{function is not differentiable}}
 @differentiable(reverse)
 // expected-note @+1 {{when differentiating this function definition}}
-func testAccessorCoroutines(_ x: HasCoroutineAccessors) -> HasCoroutineAccessors {
+func testAccessorCoroutinesRead(_ x: HasCoroutineReadAccessors) -> Float {
+  // We do not support differentiation of _read accessors
+  // expected-note @+1 {{cannot differentiate through a direct yield result}}
```
Comment (asl) [inline test/AutoDiff/SILOptimizer/activity_analysis.swift:597]:
```suggestion
  // expected-note @+1 {{cannot differentiate through a '_read' accessor}}
```

File: test/AutoDiff/SILOptimizer/differentiation_diagnostics.swift [hunk 3]
```diff
@@ -806,45 +805,56 @@ public func f_54819<Scalar: Differentiable>(
...
     _modify { yield &stored }
   }
 }
+
 // expected-error @+2 {{function is not differentiable}}
 // expected-note @+2 {{when differentiating this function definition}}
 @differentiable(reverse)
-func testAccessorCoroutines(_ x: HasCoroutineAccessors) -> HasCoroutineAccessors {
+func testReadAccessorCoroutines(_ x: HasReadAccessors) -> Float {
+  // expected-note @+1 {{cannot differentiate through a direct yield result}}
```
Comment (asl) [inline test/AutoDiff/SILOptimizer/differentiation_diagnostics.swift:829]:
```suggestion
  // expected-note @+1 {{cannot differentiate through a '_read' accessor}}
```

**Comment by asl** [comment]
@swift-ci please test

**Comment by asl** [comment]
@swift-ci please test

**Comment by asl** [comment]
@swift-ci please test macos platform

**Comment by asl** [comment]
@swift-ci please test

**Comment by asl** [comment]
@swift-ci please test macos platform

**Comment by asl** [comment]
@swift-ci please test macos platform

**Comment by asl** [comment]
@swift-ci please test

**Comment by asl** [comment]
@swift-ci please test

**Comment by asl** [comment]
@swift-ci please test

**Comment by asl** [comment]
@swift-ci please test macos platform

**Pull request merged**